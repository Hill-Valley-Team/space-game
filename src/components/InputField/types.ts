import { ChangeEvent } from 'react';
import { HookOutputFunction } from '../../hooks/useFormInput/types';

export type DefaultInputOnChangeFunction = (e: ChangeEvent<HTMLInputElement>) => void;

export type InputFieldProps = {
  id?: string;
  name?: string;
  type?: 'text' | 'email' | 'hidden' | 'number' | 'password' | 'tel';
  value?: string;
  className?: string;
  label?: string;
  errorText?: string | null;
  isValid?: boolean;
  disabled?: boolean;
  onChange?: HookOutputFunction;
};
