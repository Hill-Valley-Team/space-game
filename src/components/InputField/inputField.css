.input-field {
  position: relative;
  min-height: 50px;

  &__label {
    position: absolute;
    z-index: 0;
    bottom: 28px;
    display: block;
    font-size: 13px;
    line-height: 13px;
    color: $text-color-light;
  }

  &__input {
    position: relative;
    width: 100%;
    border: none;
    outline: none;
    border-bottom: 1px solid $primary-color;
    background-color: transparent;
    height: 35px;
    text-align: center;
  }

  &__error-text {
    color: $error-text-color;
    margin-top: 7px;
    font-size: 8px;
    line-height: 8px;
    display: block;
    text-align: left;
  }

  &_labeled {
    min-height: 62px;

    .input-field__input {
      text-align: left;
      height: 25px;
    }

    .input-field__label {
      position: relative;
      bottom: auto;
      font-size: 12px;
    }
  }

  &__password-control {
    position: absolute;
    z-index: 100;
    right: 0;
    width: 25px;
    height: 25px;
    border: none;
    cursor: pointer;

    &_view {
      background: url('./static/pass_view.svg') 50% 50% no-repeat;
    }

    &_hide {
      background: url('./static/pass_hide.svg') 50% 50% no-repeat;
    }
  }
}
