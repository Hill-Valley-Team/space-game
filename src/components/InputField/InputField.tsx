import block from 'bem-cn';
import React, { ChangeEvent, useState } from 'react';
import { InputFieldProps } from './types';
import './inputField.css';

const b = block('input-field');

export const InputField = ({
  id,
  type = 'text',
  className,
  isValid = false,
  view = 'default',
  label,
  errorText = 'Error in this field',
  onChangeHandle,
  ...props
}: InputFieldProps) => {
  const [passVisible, setPassVisible] = useState(false);

  const errorField = !isValid ? <div className={b('error-text')}>{errorText}</div> : null;

  const labelField = label ? (
    <label htmlFor={id} className={b('label')}>
      {label}
    </label>
  ) : null;

  const togglePasswordControl = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    setPassVisible(!passVisible);
  }

  const passwordControl =
    type === 'password' ? (
      <button
        type="button"
        onClick={togglePasswordControl}
        aria-label="Показать пароль"
        className={b('password-control', { [passVisible ? 'view' : 'hide']: true })}
      />
    ) : null;

  const handleChange = (e: ChangeEvent<HTMLInputElement>, equal?: string) => {
    const newValue = e.target.value;
    if (onChangeHandle) {
      return onChangeHandle({ value: newValue, equal });
    }
    return null;
  };

  return (
    <div className={b({ [view]: true }).mix(className)}>
      {labelField}
      {passwordControl}
      <input
        id={id}
        type={type === 'password' && passVisible ? 'text' : type}
        className={b('input')}
        onChange={handleChange}
        {...props}
      />
      {errorField}
    </div>
  );
}
