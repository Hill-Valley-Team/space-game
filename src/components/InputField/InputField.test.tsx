import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { InputField } from './InputField';

describe('InputField', () => {
  it('should render with default props', () => {
    expect.assertions(1);

    const input = renderer.create(<InputField />).toJSON();

    expect(input).toMatchSnapshot();
  });

  it('should render valid', () => {
    expect.assertions(1);

    const input = renderer.create(<InputField isValid />).toJSON();

    expect(input).toMatchSnapshot();
  });

  it('should render with type', () => {
    expect.assertions(1);

    const type = 'password';
    const input = renderer.create(<InputField type={type} />).toJSON();

    expect(input).toMatchSnapshot();
  });

  it('should render with label', () => {
    expect.assertions(1);

    const input = renderer.create(<InputField label="Label text" />).toJSON();

    expect(input).toMatchSnapshot();
  });

  it('should render with value', () => {
    expect.assertions(1);

    const value = 'test value';
    const input = renderer.create(<InputField value={value} />);

    expect(input).toMatchSnapshot();
  });

  it('should render with className props', () => {
    expect.assertions(1);

    const className = 'custom';
    const { container } = render(<InputField className={className} />);
    const parent = container.firstChild;

    expect(parent).toHaveClass(className);
  });

  it('should be changeable', () => {
    // expect.assertions(1);
    const newValue = 'test_value';
    const handleOnChange = jest.fn();
    const { getByRole } = render(<InputField onChangeHandle={handleOnChange} />);
    const input = getByRole('textbox');

    fireEvent.change(input, { target: { newValue } });

    expect(handleOnChange).toHaveBeenCalledTimes(1);
  });
});
