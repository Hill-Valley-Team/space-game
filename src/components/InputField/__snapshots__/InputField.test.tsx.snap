// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should render valid 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <input
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onChange={[Function]}
    type="text"
  />
</div>
`;

exports[`InputField should render with default props 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <input
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onChange={[Function]}
    type="text"
  />
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Error in this field
  </div>
</div>
`;

exports[`InputField should render with label 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <label
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Label text
  </label>
  <input
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onChange={[Function]}
    type="text"
  />
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Error in this field
  </div>
</div>
`;

exports[`InputField should render with type 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <input
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onChange={[Function]}
    type="password"
  />
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Error in this field
  </div>
</div>
`;

exports[`InputField should render with value 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <input
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    onChange={[Function]}
    type="text"
    value="test value"
  />
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    Error in this field
  </div>
</div>
`;
